### DISCUSSION TITLE
Improving Documentation of BokehJS

### DISCUSSION THREAD

**Discussion by matteosox** [bokeh/bokeh#14069] [Documentation]
I find Bokeh to be the best, most powerful Python-based tool for data visualization, but I struggle to (1) make those visualizations as interactive as I'd like with CustomJS callbacks since the JS API doesn't seem to have any documentation, and (2) export those visualizations outside the typical notebook environment, e.g. sharing with others on the web or even embedding them into a website.

I'd love to make progress on improving bokeh's functionality in these areas, and probably to start would be most useful in simply documenting existing functionality.

**Comment by bryevdv** [comment]
Hi @matteosox thanks for chiming in!

> make those visualizations as interactive as I'd like with CustomJS callbacks since the JS API doesn't seem to have any documentation

Just clarifying that the context here is specifically for `CustomJS` callback code, correct? It is true that the BokehJS API does not have much documentation, but it's also true that in the context of `CustomJS` code, very little of the BokehJS API actually comes into play, except for updating models and their properties, and models and properties are 1-1 identical with the Python side (e.g. updating `source.data` or setting `row.children` there is no difference really between Python and `CustomJS` usage).

So some ideas that do come to mind:

* Reinforcing that that Python reference guide for models is actually also the best/appropriate place for seeing what properties exist to be updated in `CustomJS` code as well
* Making it clear that BokehJS can only "automagically" respond to full assignments (i.e. it can detect `source.data = new_data` but not `source.data["some_col"].push(...)`

Are there other specific things you can point to @matteosox that you have run into in the past that might be good to add to this list?

> export those visualizations outside the typical notebook environment, e.g. sharing with others on the web or even embedding them into a website.

I think more context is needed here. There's an existing chapter on exporting/embedding at https://docs.bokeh.org/en/latest/docs/user_guide/output/embed.html so the first thing to understand is what task you wanted to accomplish that was not covered by the cases there.

**Reply by bjfar** [reply]
Here's one: I have struggled a lot to understand what happens with CDSViews and Filters at the JS level. I have often wanted to e.g. use a filtered view into the data in some customJS, i.e. be able to tell which indices of the source data would be selected by the filter, but I have failed miserably. And since operations with CDSViews and Filters are things that Bokeh mostly manages under the hood there is very little in the Python documentation that gives me clues as to how I can make use of them in custom JS.
